### ISSUE TITLE
Inline constraint/objective definitions in Style

### ISSUE THREAD

**Issue by keenancrane** [penrose/penrose#307]
We've talked for a while now about user extensibility of functions, objectives, and constraints (which currently come from a fixed list of definitions within Penrose itself).  I propose that such extensibility is supported within Style itself.

The most basic example is just a simple computation that gets re-used many times within a Style, but is very specific to that Style (and hence should _not_ be part of the fixed list of functions available in Penrose).  For instance, let's say that for the Style I'm developing a bunch of different blocks need to be able to compute Milnor's Lobachevsky function.  Then somewhere at the top of my Style file, I want to be able to define something like

```
milnorLobachevsky(x) {
   -- do something involving Chebyshev polynomials...
   return y
}
```

(not necessarily with this syntax).

---

In terms of objectives/constraints, as a running example, suppose I want to construct this figure:

![image](https://user-images.githubusercontent.com/3604525/71117747-bf88fb00-21a4-11ea-9453-2549a80b298a.png)

A natural specification would be something like:

```
ensure Intersect( circle1, circle2 )
ensure PointOnCircle( p, circle1 )
ensure PointOnCircle( p, circle2 )
```

I.e., there are two circles that must intersect, and a point p that must be on the intersection of the two circles.  Suppose that the predicate `PointOnCircle` is not already defined in Penrose.  Then a good syntax for its definition in Style would be something like

```
PointOnCircle( p, C ) := Norm(p.position - C.center) - C.radius
```

In other words, we define a function that evaluates to zero if and only if the desired predicate is satisfied (in this case: the distance between the point and the circle's center is equal to the circle's radius).  This function has global scope and can be used anywhere in the Style program, or in any file that cascades on this Style program.

**Comment by hypotext** [comment]
See this library for one approach to defining constraints on the fly: https://github.com/soney/constraintjs/wiki/Constraint-Variables

**Comment by wodeni** [comment]
See this library for allowing operator overloading in JS: http://paperjs.org/tutorials/geometry/mathematical-operations/#mathematical-operations.

**Comment by hypotext** [comment]
Note: we are going to need to be able to differentiate through all inline computations.

**Comment by keenancrane** [comment]
Note that we will need to be able to differentiate through user-defined functions!

**Comment by maxkrieger** [comment]
Something I keep forgetting/mixing up is how Stella's sugared-TS compares to inline defs in style AST. Are both necessary or at fundamentally different layers? I'm of course partial to unifying everything in the same grammar.

**Comment by hypotext** [comment]
@maxkrieger Stella's (@strout18) sugared-TS is at the system layer, but the same approach could (and should) work well with inline defs in Style. It would be even better if functions related to a specific Style could appear in that Style, rather than being added to our whole system.

**Comment by hypotext** [comment]
Another suggestion is to allow people to define anonymous constraints/objectives with some set of syntax, e.g. writing “equals(A, B)” inline in Style as “A == B” instead.

Suggested by P3 in domain expert study on 6/11/21
